packages/init/node-async_hooks.js

```diff
@@ -0,0 +1,229 @@
+      if (setAsyncSymbol(description, sym)) {
+        found += 1;
+      }
+    }
+  });
+  return found;
+};
+
+// To get the `destroyed` symbol installed on promises by async_hooks,
+// the only option is to create and enable an AsyncHook.
+// Different versions of Node handle this in various ways.
+const findAsyncSymbolsFromPromiseCreateHook = () => {
+  const bootstrapData = [];
+
+  {
+    const bootstrapHook = createHook({
+      init(asyncId, type, triggerAsyncId, resource) {
+        if (type !== 'PROMISE') return;
+        // process._rawDebug('Bootstrap', asyncId, triggerAsyncId, resource);
+        bootstrapData.push({ asyncId, triggerAsyncId, resource });
+      },
+      destroy(_asyncId) {
+        // Needs to be present to trigger the addition of the destroyed symbol
+      },
+    });
+
+    bootstrapHook.enable();
+    // Use a never resolving promise to avoid triggering settlement hooks
+    const trigger = new Promise(() => {});
+    bootstrapHook.disable();
+
+    // In some versions of Node, async_hooks don't give access to the resource
+    // itself, but to a "wrapper" which is basically hooks metadata for the promise
+    const promisesData = bootstrapData.filter(
+      ({ resource }) => Promise.resolve(resource) === resource,
+    );
+    bootstrapData.length = 0;
+    const { length } = promisesData;
+    if (length > 1) {
+      // process._rawDebug('Found multiple potential candidates');
+    }
+
+    const promiseData = promisesData.find(
+      ({ resource }) => resource === trigger,
+    );
+    if (promiseData) {
+      bootstrapData.push(promiseData);
+    } else if (length) {
+      // process._rawDebug('No candidates matched');
+    }
+  }
+
+  if (bootstrapData.length) {
+    // Normally all promise hooks are disabled in a subsequent microtask
+    // That means Node versions that modify promises at init will still
+    // trigger our proto hooks for promises created in this turn
+    // The following trick will disable the internal promise init hook
+    // However, only do this for destroy modifying versions, since some versions
+    // only modify promises if no destroy hook is requested, and do not correctly
+    // reset the internal init promise hook in those case. (e.g. v14.16.2)
+    const resetHook = createHook({});
+    resetHook.enable();
+    resetHook.disable();
+
+    const { asyncId, triggerAsyncId, resource } = bootstrapData.pop();
+    const symbols = Object.getOwnPropertySymbols(resource);
+    // const { length } = symbols;
+    let found = 0;
+    // if (length !== 3) {
+    //   process._rawDebug(`Found ${length} symbols on promise:`, ...symbols);
+    // }
+    symbols.forEach(symbol => {
+      const value = resource[symbol];
+      let type;
+      if (value === asyncId) {
+        type = 'async_id_symbol';
+      } else if (value === triggerAsyncId) {
+        type = 'trigger_async_id_symbol';
+      } else if (typeof value === 'object' && 'destroyed' in value) {
```

Comment:
Should this be an own check?
```suggestion
      } else if (typeof value === 'object' && Reflect.getOwnPropertyDescriptor(value, 'destroyed') !== undefined) {
```

Reply:
I think that's overkill